RecoCTPPS/ProtonReconstruction/plugins/CTPPSProtonReconstruction.cc

```diff
@@ -0,0 +1,200 @@
+
+#include "CondFormats/RunInfo/interface/LHCInfo.h"
+#include "CondFormats/DataRecord/interface/LHCInfoRcd.h"
+
+#include "CondFormats/DataRecord/interface/CTPPSOpticsRcd.h"
+#include "CondFormats/CTPPSReadoutObjects/interface/LHCOpticalFunctionsCollection.h"
+
+//----------------------------------------------------------------------------------------------------
+
+class CTPPSProtonReconstruction : public edm::stream::EDProducer<>
+{
+  public:
+    explicit CTPPSProtonReconstruction(const edm::ParameterSet&);
+    ~CTPPSProtonReconstruction() {}
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  private:
+    virtual void produce(edm::Event&, const edm::EventSetup&) override;
+
+    edm::EDGetTokenT< std::vector<CTPPSLocalTrackLite> > tracksToken_;
+
+    unsigned int verbosity_;
+
+    bool doSingleRPReconstruction_;
+    bool doMultiRPReconstruction_;
+
+    ProtonReconstructionAlgorithm algorithm_;
+
+    edm::ESWatcher<LHCInfoRcd> lhcInfoWatcher_;
+    float currentCrossingAngle_;
+
+    std::unordered_map<unsigned int, LHCOpticalFunctionsSet> opticalFunctions_;
+
+    const std::string singleRPLabel = "singleRP";
+    const std::string multiRPLabel = "multiRP";
+};
+
+//----------------------------------------------------------------------------------------------------
+//----------------------------------------------------------------------------------------------------
+
+using namespace std;
+using namespace edm;
+
+//----------------------------------------------------------------------------------------------------
+
+CTPPSProtonReconstruction::CTPPSProtonReconstruction(const edm::ParameterSet& iConfig) :
+  tracksToken_(consumes< std::vector<CTPPSLocalTrackLite> >(iConfig.getParameter<edm::InputTag>("tagLocalTrackLite"))),
+  verbosity_(iConfig.getUntrackedParameter<unsigned int>("verbosity", 0)),
+  doSingleRPReconstruction_(iConfig.getParameter<bool>("doSingleRPReconstruction")),
+  doMultiRPReconstruction_(iConfig.getParameter<bool>("doMultiRPReconstruction")),
+  algorithm_(iConfig.getParameter<bool>("fitVtxY"), verbosity_),
+  currentCrossingAngle_(-1.)
+{
+  if (doSingleRPReconstruction_)
+    produces<vector<reco::ProtonTrack>>(singleRPLabel);
+
+  if (doMultiRPReconstruction_)
+    produces<vector<reco::ProtonTrack>>(multiRPLabel);
+}
+
+//----------------------------------------------------------------------------------------------------
+
+void CTPPSProtonReconstruction::fillDescriptions(ConfigurationDescriptions& descriptions)
+{
+  ParameterSetDescription desc;
+  desc.setUnknown();
+  descriptions.addDefault(desc);
+}
+
+//----------------------------------------------------------------------------------------------------
+
+void CTPPSProtonReconstruction::produce(Event& event, const EventSetup &eventSetup)
+{
+  if (verbosity_)
+    printf("\n---------- %u:%llu ----------\n", event.id().run(), event.id().event());
```

Comment:
We need to drop these `printf`/`verbosity` flags at some point and move to [the `LogXXX`](https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideMessageLogger) capability of the framework (which is by the way thread-safe... unless we want to deal with `mutex`es...)